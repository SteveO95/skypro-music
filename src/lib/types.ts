export enum FilterType {
	author = 'исполнителю',
	year = 'году выпуска',
	genre = 'жанру',
}

export type TrackDataType = {
	_id: number
	name: string
	author: string
	release_date: string
	genre: string[]
	duration_in_seconds: number
	album: string
	logo: { type: string; data: unknown[] }
	track_file: string
	staredUser: number[]
}

export type UserDataType = {
	email: string
	username: string
	_id: number
}

export type TokenType = {
	access: string
	refresh: string
}

export type SignInDataType = {
	email: string
	password: string
}

export type SignUpDataType = {
	email: string
	password: string
	username: string
}
