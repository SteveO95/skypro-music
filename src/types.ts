// Структура Трека
export type trackType = {
	dispatch(arg0: void): unknown;
	id: number;
	name: string;
	author: string;
	release_date: string;
	genre: string;
	duration_in_seconds: number;
	album: string;
	logo: null | string;
	track_file: string;
	stared_user: Array<userType>;
};

type userType = {
	id: number;
	username: string;
	first_name: string;
	last_name: string;
	email: string;
};
export type getTrackResponse = {
	error: string | null;
	data: trackType[] | null;
};
