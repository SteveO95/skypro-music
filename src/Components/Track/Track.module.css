.playlist__item {
	width: 100%;
	display: block;
	margin-bottom: 12px;
	cursor: pointer;
}

.playlist__track {
	display: -webkit-box;
	display: -ms-flexbox;
	display: flex;
	-webkit-box-orient: horizontal;
	-webkit-box-direction: normal;
	-ms-flex-direction: row;
	flex-direction: row;
	-webkit-box-pack: justify;
	-ms-flex-pack: justify;
	justify-content: space-between;
	-webkit-box-align: center;
	-ms-flex-align: center;
	align-items: center;
}

.track__title {
	display: -webkit-box;
	display: -ms-flexbox;
	display: flex;
	-webkit-box-orient: horizontal;
	-webkit-box-direction: normal;
	-ms-flex-direction: row;
	flex-direction: row;
	-webkit-box-align: center;
	-ms-flex-align: center;
	align-items: center;
	width: 447px;
}

.track__title_image {
	width: 51px;
	height: 51px;
	padding: 16px;
	background: #313131;
	display: -webkit-box;
	display: -ms-flexbox;
	display: flex;
	-webkit-box-align: center;
	-ms-flex-align: center;
	align-items: center;
	-webkit-box-pack: center;
	-ms-flex-pack: center;
	justify-content: center;
	margin-right: 17px;
}

.track__title_svg {
	width: 18px;
	height: 17px;
	fill: transparent;
	stroke: #4e4e4e;
}

.track__title_link {
	font-style: normal;
	font-weight: 400;
	font-size: 16px;
	line-height: 24px;
	color: #ffffff;
}

.track__title_span {
	font-style: normal;
	font-weight: 400;
	font-size: 16px;
	line-height: 24px;
	color: #4e4e4e;
}

.track__author {
	width: 321px;
	display: -webkit-box;
	display: -ms-flexbox;
	display: flex;
	-webkit-box-pack: start;
	-ms-flex-pack: start;
	justify-content: flex-start;
}

.track__author_link {
	font-style: normal;
	font-weight: 400;
	font-size: 16px;
	line-height: 24px;
	color: #ffffff;
	text-align: left;
}

.track__album {
	width: 245px;
}

.track__album_link {
	font-style: normal;
	font-weight: 400;
	font-size: 16px;
	line-height: 24px;
	color: #696969;
}

.track__time {
	display: flex;
	align-items: center;
}

.track__time_svg {
	width: 14px;
	height: 12px;
	margin-right: 17px;
	fill: transparent;
	stroke: #696969;
}

.track__time_text {
	font-style: normal;
	font-weight: 400;
	font-size: 16px;
	line-height: 24px;
	text-align: right;
	color: #696969;
}

.playing_dot {
	width: 16px;
	height: 16px;
	background-color: #b672ff;
	border-radius: 8px;
	display: block;
	animation: bubble_out 0.6s ease-in-out infinite both;
}
.not_playing_dot {
	width: 16px;
	height: 16px;
	background-color: #b672ff;
	border-radius: 8px;
	display: block;
}

.like_heartbeat_animation {
	animation: heartbeat 1.7s ease-in-out both;
}
.dislike_jello_animation {
	animation: jello-horizontal 1.2s both;
}

@keyframes bubble_out {
	0%,
	to {
		transform: scale(0.5);
	}
	50% {
		transform: scale(1);
	}
}

@keyframes jello-horizontal {
	0% {
		-webkit-transform: scale3d(1, 1, 1);
		transform: scale3d(1, 1, 1);
	}
	30% {
		-webkit-transform: scale3d(1.25, 0.75, 1);
		transform: scale3d(1.25, 0.75, 1);
	}
	40% {
		-webkit-transform: scale3d(0.75, 1.25, 1);
		transform: scale3d(0.75, 1.25, 1);
	}
	50% {
		-webkit-transform: scale3d(1.15, 0.85, 1);
		transform: scale3d(1.15, 0.85, 1);
	}
	65% {
		-webkit-transform: scale3d(0.95, 1.05, 1);
		transform: scale3d(0.95, 1.05, 1);
	}
	75% {
		-webkit-transform: scale3d(1.05, 0.95, 1);
		transform: scale3d(1.05, 0.95, 1);
	}
	100% {
		-webkit-transform: scale3d(1, 1, 1);
		transform: scale3d(1, 1, 1);
	}
}

@keyframes heartbeat {
	from {
		-webkit-transform: scale(1);
		transform: scale(1);
		-webkit-transform-origin: center center;
		transform-origin: center center;
		-webkit-animation-timing-function: ease-out;
		animation-timing-function: ease-out;
	}
	10% {
		-webkit-transform: scale(0.91);
		transform: scale(0.91);
		-webkit-animation-timing-function: ease-in;
		animation-timing-function: ease-in;
	}
	17% {
		-webkit-transform: scale(0.98);
		transform: scale(0.98);
		-webkit-animation-timing-function: ease-out;
		animation-timing-function: ease-out;
	}
	33% {
		-webkit-transform: scale(0.87);
		transform: scale(0.87);
		-webkit-animation-timing-function: ease-in;
		animation-timing-function: ease-in;
	}
	45% {
		-webkit-transform: scale(1);
		transform: scale(1);
		-webkit-animation-timing-function: ease-out;
		animation-timing-function: ease-out;
	}
}
