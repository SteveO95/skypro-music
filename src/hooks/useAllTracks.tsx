import { getAllTracks } from "@/services/api";
import { TrackType } from "@/types/tracks";
import { useEffect, useState } from "react";

const useAllTracks = (): [TrackType[], boolean, boolean] => {
  const [tracks, setTracks] = useState<TrackType[]>([]);
  const [loading, setLoading] = useState<boolean>(true);
  const [error, setError] = useState<boolean>(false);

  useEffect(() => {
    const fetchAllTracks = async () => {
      try {
        const data = await getAllTracks();
        setTracks(data);
      } catch (err) {
        setError(true);
      } finally {
        setLoading(false);
      }
    };

    fetchAllTracks();
  }, []);

  return [tracks, loading, error];
};

export default useAllTracks;
