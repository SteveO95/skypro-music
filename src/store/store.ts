import {
	TypedUseSelectorHook,
	useDispatch,
	useSelector,
	useStore,
} from 'react-redux'
import { combineReducers, configureStore } from '@reduxjs/toolkit'
import { userReducer } from './features/userSlice'
import { playlistReducer } from './features/playlistSlice'

export const makeStore = () => {
	return configureStore({
		reducer: combineReducers({
			user: userReducer,
			playlist: playlistReducer,
		}),
	})
}

export type AppStore = ReturnType<typeof makeStore>
export type RootState = ReturnType<AppStore['getState']>
export type AppDispatch = AppStore['dispatch']

export const useAppDispatch: () => AppDispatch = useDispatch
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector
export const useAppStore: () => AppStore = useStore
