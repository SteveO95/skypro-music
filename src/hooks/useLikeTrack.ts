import { tracksApi } from '@/api/tracksApi'
import { TrackDataType } from '@/lib/types'
import { setDislike, setLike } from '@/store/features/playlistSlice'
import { useAppDispatch, useAppSelector } from '@/store/store'

export function useLikeTrack(track: TrackDataType) {
	const dispatch = useAppDispatch()
	const { user, tokens } = useAppSelector(state => state.user)
	const { favoritePlaylist } = useAppSelector(state => state.playlist)

	const isLiked = !!favoritePlaylist.find(el => el._id === track._id)

	const handleLike = async (event: React.MouseEvent<HTMLDivElement>) => {
		event.stopPropagation()

		if (!user || !tokens) {
			alert('Ставить лайки могут только авторизованные пользователи')
			return
		}

		const fetchAction = isLiked
			? tracksApi.deleteFavoriteTrack
			: tracksApi.addFavoriteTrack

		const storeAction = isLiked ? setDislike : setLike

		try {
			await fetchAction({
				trackId: track._id,
				access: tokens.access,
				refresh: tokens.refresh,
			})
			dispatch(storeAction(track))
		} catch (err) {
			const error = err as Error
			console.error(error.message)
		}
	}

	return {
		isLiked,
		handleLike,
	}
}
