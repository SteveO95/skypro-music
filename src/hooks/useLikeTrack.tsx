import { handleError } from "@/components/Toast/Toast";
import { dislikeTrack, likeTrack } from "@/services/api";
import { setDislikeTrack, setLikeTrack } from "@/store/features/trackSlice";
import { useAppDispatch, useAppSelector } from "@/store/store";

const useLikeTrack = (trackID: number) => {
  const dispatch = useAppDispatch();

  const tokens = {
    access: useAppSelector((state) => state.auth.accessToken),
    refresh: useAppSelector((state) => state.auth.refreshToken),
  };
  const likedTraks = useAppSelector((state) => state.track.likedPlaylistState);

  // Получаем состояние лайка из избранных треков
  // или получать в качестве пропса
  const isLiked = !!likedTraks ? !!likedTraks.find((track) => track.id === trackID) : false;

  const handleLike = async (e: React.MouseEvent) => {
    e.stopPropagation();
    if (!tokens.access || !tokens.refresh) return handleError("Вы не авторизованы");

    const action = isLiked ? dislikeTrack : likeTrack;

    try {
      await action({
        trackId: trackID,
        access: tokens.access,
        refresh: tokens.refresh,
      });

      if (isLiked) {
        dispatch(setDislikeTrack(trackID));
      } else {
        dispatch(setLikeTrack(trackID));
      }
    } catch (error) {
      handleError("Непредвиденная ошибка");
      console.error(error);
    }
  };

  return { handleLike, isLiked };
};

export default useLikeTrack;
