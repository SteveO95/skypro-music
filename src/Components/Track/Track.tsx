import { useLikeTrack } from '@/hooks/useLikeTrack'
import { TrackDataType } from '@/lib/types'
import { setCurrTrack, setIsPlaying } from '@/store/features/playlistSlice'
import { useAppDispatch, useAppSelector } from '@/store/store'
import { trackFormattedTime } from '@/utils/helpers'
import classNames from 'classnames'
import styles from './Track.module.css'

export default function Track(track: TrackDataType) {
	const dispatch = useAppDispatch()
	const { mainPlaylist, currTrack, isPlaying } = useAppSelector(
		state => state.playlist,
	)

	const { isLiked, handleLike } = useLikeTrack(track)

	const handlePlay = (track: TrackDataType, tracks: TrackDataType[]) => {
		dispatch(setCurrTrack({ currTrack: track, mainPlaylist: tracks }))
		dispatch(setIsPlaying(true))
	}

	return (
		<div
			className={styles.playlist__item}
			onClick={() => handlePlay(track, mainPlaylist)}
		>
			<div className={styles.playlist__track}>
				<div className={styles.track__title}>
					<div className={styles.track__title_image}>
						{currTrack?._id === track._id ? (
							<div
								className={
									isPlaying ? styles.playing_dot : styles.not_playing_dot
								}
							></div>
						) : (
							<svg className={styles.track__title_svg}>
								<use xlinkHref='/img/icon/sprite.svg#icon-note' />
							</svg>
						)}
					</div>
					<div>
						<a className={styles.track__title_link} href='#'>
							{track.name} <span className={styles.track__title_span} />
						</a>
					</div>
				</div>
				<div className={styles.track__author}>
					<a className={styles.track__author_link} href='#'>
						{track.author}
					</a>
				</div>
				<div className={styles.track__album}>
					<a className={styles.track__album_link} href='#'>
						{track.album}
					</a>
				</div>
				<div className={styles.track__time}>
					<div onClick={handleLike}>
						<svg
							className={classNames(
								styles.track__time_svg,
								isLiked
									? styles.like_heartbeat_animation
									: styles.dislike_jello_animation,
							)}
						>
							<use
								xlinkHref={`/img/icon/sprite.svg#${isLiked ? 'icon-like' : 'icon-dislike'}`}
							/>
						</svg>
					</div>
					<span className={styles.track__time_text}>
						{trackFormattedTime(track.duration_in_seconds)}
					</span>
				</div>
			</div>
		</div>
	)
}
