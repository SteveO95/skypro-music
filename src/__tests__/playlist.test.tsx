import Track from '@/components/Track/Track'
import { TrackDataType } from '@/lib/types'
import { useAppDispatch, useAppSelector } from '@/store/store'
import { useLikeTrack } from '@/hooks/useLikeTrack'
import '@testing-library/jest-dom'
import renderer from 'react-test-renderer'
import { cleanup } from '@testing-library/react'

jest.mock('../store/store.ts')
jest.mock('../hooks/useLikeTrack.ts')

const mockDispatch = useAppDispatch as jest.Mock
const mockSelector = useAppSelector as jest.Mock
const mockLike = useLikeTrack as jest.Mock

describe('Плейлист', () => {
	beforeEach(() => {
		mockDispatch.mockReturnValue(jest.fn())
		mockSelector.mockReturnValue(jest.fn())
		mockLike.mockReturnValue(jest.fn())
	})
	afterEach(() => {
		cleanup()
	})

	const trackListTests: TrackDataType[] = [
		{
			_id: 1,
			name: 'test 1',
			author: 'test 1',
			album: 'test 1',
			duration_in_seconds: 100,
			track_file: '',
			release_date: '2021-01-01',
			genre: [],
			logo: { type: '', data: [] },
			staredUser: [],
		},
		{
			_id: 2,
			name: 'test 2',
			author: 'test 2',
			album: 'test 2',
			duration_in_seconds: 200,
			track_file: '',
			release_date: '2022-02-02',
			genre: [],
			logo: { type: '', data: [] },
			staredUser: [],
		},
		{
			_id: 3,
			name: 'test 3',
			author: 'test 3',
			album: 'test 3',
			duration_in_seconds: 300,
			track_file: '',
			release_date: '2023-03-03',
			genre: [],
			logo: { type: '', data: [] },
			staredUser: [],
		},
	]

	trackListTests.forEach(track => {
		it('рендер списка треков', () => {
			const tracks = renderer
				.create(<Track key={track._id} {...track} />)
				.toJSON()
			expect(tracks).toMatchSnapshot()
		})
	})
})
