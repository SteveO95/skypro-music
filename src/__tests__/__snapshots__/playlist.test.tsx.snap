// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Плейлист рендер списка треков 1`] = `
<div
  className="playlist__item"
  onClick={[Function]}
>
  <div
    className="playlist__track"
  >
    <div
      className="track__title"
    >
      <div
        className="track__title_image"
      >
        <svg
          className="track__title_svg"
        >
          <use
            xlinkHref="/img/icon/sprite.svg#icon-note"
          />
        </svg>
      </div>
      <div>
        <a
          className="track__title_link"
          href="#"
        >
          test 1
           
          <span
            className="track__title_span"
          />
        </a>
      </div>
    </div>
    <div
      className="track__author"
    >
      <a
        className="track__author_link"
        href="#"
      >
        test 1
      </a>
    </div>
    <div
      className="track__album"
    >
      <a
        className="track__album_link"
        href="#"
      >
        test 1
      </a>
    </div>
    <div
      className="track__time"
    >
      <div>
        <svg
          className="track__time_svg dislike_jello_animation"
        >
          <use
            xlinkHref="/img/icon/sprite.svg#icon-dislike"
          />
        </svg>
      </div>
      <span
        className="track__time_text"
      >
        1:40
      </span>
    </div>
  </div>
</div>
`;

exports[`Плейлист рендер списка треков 2`] = `
<div
  className="playlist__item"
  onClick={[Function]}
>
  <div
    className="playlist__track"
  >
    <div
      className="track__title"
    >
      <div
        className="track__title_image"
      >
        <svg
          className="track__title_svg"
        >
          <use
            xlinkHref="/img/icon/sprite.svg#icon-note"
          />
        </svg>
      </div>
      <div>
        <a
          className="track__title_link"
          href="#"
        >
          test 2
           
          <span
            className="track__title_span"
          />
        </a>
      </div>
    </div>
    <div
      className="track__author"
    >
      <a
        className="track__author_link"
        href="#"
      >
        test 2
      </a>
    </div>
    <div
      className="track__album"
    >
      <a
        className="track__album_link"
        href="#"
      >
        test 2
      </a>
    </div>
    <div
      className="track__time"
    >
      <div>
        <svg
          className="track__time_svg dislike_jello_animation"
        >
          <use
            xlinkHref="/img/icon/sprite.svg#icon-dislike"
          />
        </svg>
      </div>
      <span
        className="track__time_text"
      >
        3:20
      </span>
    </div>
  </div>
</div>
`;

exports[`Плейлист рендер списка треков 3`] = `
<div
  className="playlist__item"
  onClick={[Function]}
>
  <div
    className="playlist__track"
  >
    <div
      className="track__title"
    >
      <div
        className="track__title_image"
      >
        <svg
          className="track__title_svg"
        >
          <use
            xlinkHref="/img/icon/sprite.svg#icon-note"
          />
        </svg>
      </div>
      <div>
        <a
          className="track__title_link"
          href="#"
        >
          test 3
           
          <span
            className="track__title_span"
          />
        </a>
      </div>
    </div>
    <div
      className="track__author"
    >
      <a
        className="track__author_link"
        href="#"
      >
        test 3
      </a>
    </div>
    <div
      className="track__album"
    >
      <a
        className="track__album_link"
        href="#"
      >
        test 3
      </a>
    </div>
    <div
      className="track__time"
    >
      <div>
        <svg
          className="track__time_svg dislike_jello_animation"
        >
          <use
            xlinkHref="/img/icon/sprite.svg#icon-dislike"
          />
        </svg>
      </div>
      <span
        className="track__time_text"
      >
        5:00
      </span>
    </div>
  </div>
</div>
`;
